File: te_phoenix/lib/te_phoenix/battle/loot.ex
defmodule TePhoenix.Battle.Loot do
  @moduledoc """
  Loot system: Steal, Drop Tables, Overkill bonus, Battle Chain, Battle Rating.
  Ported from te/battle/loot.js
  require Logger
  """

  alias TePhoenix.Repo
  import Ecto.Query

  # ══════════════════════════════════════════════════════════════════
  # STEAL
  # ══════════════════════════════════════════════════════════════════

  @doc """
  Mid-battle steal action. Roll against target's steal table.
  Two pools: common (high chance) and rare (low chance).
  Chance = (luck + speed*0.5) vs target level.
  """
  def resolve_steal(state, actor, target, result) do
    settings = state.settings

    cond do
      not settings[:enable_steal] ->
        {state, %{result | log: ["Steal is not available." | result.log]}}

      Map.get(target, :already_stolen, false) ->
        {state, %{result | log: ["#{target.name} has nothing left to steal!" | result.log]}}

      actor.team_id == target.team_id ->
        {state, %{result | log: ["#{actor.name} can't steal from an ally!" | result.log]}}

      true ->
        steal_table = load_steal_table(target.char_id)

        if steal_table == nil or steal_table == [] do
          {state, %{result | log: ["#{target.name} has nothing to steal." | result.log]}}
        else
          base_chance = settings[:steal_base_chance] || 50
          chance = min(95, max(10,
            base_chance + (actor.luck || 0) * 2 + (actor.speed || 0) * 0.5 - (target.level || 1) * 3
          ))

          if :rand.uniform(100) > chance do
            result = %{result |
              log: ["#{actor.name} tries to steal from #{target.name}... but fails! (#{round(chance)}% chance)" | result.log],
              actions: [%{type: :steal, actor: actor.name, target: target.name, success: false} | result.actions]
            }
            {state, result}
          else
            {stolen, is_rare} = roll_steal_item(steal_table, settings)

            if stolen == nil do
              {state, %{result | log: ["#{actor.name} rummages through #{target.name}'s pockets... nothing!" | result.log]}}
            else
              # Mark target as stolen from
              target = Map.put(target, :already_stolen, true)
              state = %{state | combatants: Map.put(state.combatants, target.char_id, target)}

              # Give item to actor (DB side)
              item_info = give_stolen_item(actor.char_id, stolen)

              result = %{result |
                log: ["#{item_info.icon} #{actor.name} steals #{item_info.name}#{if stolen["qty"] > 1, do: " x#{stolen["qty"]}", else: ""} from #{target.name}!" | result.log],
                actions: [%{type: :steal, actor: actor.name, target: target.name, success: true,
                  item: %{id: stolen["item_id"], name: item_info.name, icon: item_info.icon, qty: stolen["qty"] || 1, rare: is_rare}} | result.actions]
              }
              {state, result}
            end
          end
        end
    end
  end

  defp load_steal_table(char_id) do
    case Repo.one(from n in "game_npcs", where: n.char_id == ^char_id, select: n.steal_table_json, limit: 1) do
      nil -> []
      json when is_binary(json) ->
        case Jason.decode(json) do
          {:ok, list} when is_list(list) -> list
          _ -> []
        end
      _ -> []
    end
  rescue
    _ -> []
  end

  defp roll_steal_item(steal_table, settings) do
    rare_items = Enum.filter(steal_table, fn e -> e["rarity"] == "rare" end)
    common_items = Enum.filter(steal_table, fn e -> e["rarity"] != "rare" end)
    rare_chance = settings[:steal_rare_chance] || 0.20

    cond do
      rare_items != [] and :rand.uniform() < rare_chance ->
        {Enum.random(rare_items), true}
      common_items != [] ->
        {Enum.random(common_items), false}
      true ->
        {nil, false}
    end
  end

  defp give_stolen_item(char_id, stolen) do
    item_id = stolen["item_id"]
    qty = stolen["qty"] || 1

    try do
      Repo.query!(
        "INSERT INTO character_items (character_id, item_id, quantity) VALUES (?,?,?) ON DUPLICATE KEY UPDATE quantity=quantity+VALUES(quantity)",
        [char_id, item_id, qty]
      )
    rescue
      _ -> nil
    end

    # Get item display info
    case Repo.one(from i in "game_items", where: i.id == ^item_id, select: %{name: i.name, icon: i.icon}) do
      nil -> %{name: "Unknown Item", icon: "📦"}
      info -> %{name: info.name, icon: info.icon || "📦"}
    end
  rescue
    _ -> %{name: "Unknown Item", icon: "📦"}
  end

  # ══════════════════════════════════════════════════════════════════
  # ITEM RESOLUTION (Full battle item usage)
  # ══════════════════════════════════════════════════════════════════

  @doc """
  Resolve item usage in battle. Supports:
  - Heal HP/MP
  - Throw damage (BG3-style: any item can be thrown)
  - Apply/cure status effects
  - Create terrain surfaces
  - Heal specific limbs
  """
  def resolve_item(state, actor, target, item_id, result) do
    # Check wound prevents item use
    if Map.get(actor, :wound_flags, %{})["cant_use_items"] do
      {state, %{result | log: ["#{actor.name}'s arm is too wounded to use items!" | result.log]}}
    else
      case load_item(item_id) do
        nil ->
          {state, %{result | log: ["#{actor.name} can't use that in battle!" | result.log]}}

        item ->
          effects = parse_json(item.effects)
          resolve_item_effects(state, actor, target, item, effects, result)
      end
    end
  end

  defp resolve_item_effects(state, actor, target, item, effects, result) do
    # Determine target
    throw_target = effects["throw_target"] || "self"
    item_target = case throw_target do
      "self" -> actor
      "enemy" when target != nil -> target
      "ally" when target != nil -> target
      "any" when target != nil -> target
      _ -> actor
    end

    throw_verb = if item_target.char_id == actor.char_id, do: "uses", else: "throws"
    on_text = if item_target.char_id == actor.char_id, do: "", else: " on #{item_target.name}"
    result = %{result | log: ["#{actor.name} #{throw_verb} #{item.icon || "🧪"} #{item.name}#{on_text}!" | result.log]}

    # Heal HP
    {item_target, result} =
      if effects["heal_hp"] do
        heal = effects["heal_hp"]["formula"] || "50"
        vars = TePhoenix.Battle.Combatant.formula_vars(actor, item_target)
        amount = trunc(TePhoenix.Battle.Formula.evaluate(heal, vars))
        t = TePhoenix.Battle.Combatant.apply_healing(item_target, amount)
        {t, %{result |
          log: ["#{t.name} recovers #{amount} HP!" | result.log],
          actions: [%{type: :heal, target: t.name, amount: amount} | result.actions]
        }}
      else
        {item_target, result}
      end

    # Heal MP
    {item_target, result} =
      if effects["heal_mp"] do
        heal = effects["heal_mp"]["formula"] || "30"
        vars = TePhoenix.Battle.Combatant.formula_vars(actor, item_target)
        amount = trunc(TePhoenix.Battle.Formula.evaluate(heal, vars))
        new_mp = min(item_target.max_mp, item_target.current_mp + amount)
        t = %{item_target | current_mp: new_mp}
        {t, %{result | log: ["#{t.name} recovers #{amount} MP!" | result.log]}}
      else
        {item_target, result}
      end

    # Throw damage (bombs, acid flasks)
    {item_target, result} =
      if effects["throw_damage"] do
        vars = TePhoenix.Battle.Combatant.formula_vars(actor, item_target)
        dmg = max(1, trunc(TePhoenix.Battle.Formula.evaluate(effects["throw_damage"]["formula"] || "20", vars)))
        t = TePhoenix.Battle.Combatant.apply_damage(item_target, dmg)
        {t, %{result |
          log: ["#{t.name} takes #{dmg} damage!" | result.log],
          actions: [%{type: :item_damage, target: t.name, amount: dmg, item: item.name} | result.actions]
        }}
      else
        {item_target, result}
      end

    # Surface creation (BG3-style)
    state =
      if effects["surface_on_throw"] && item_target.char_id != actor.char_id do
        key = "#{item_target.grid_x},#{item_target.grid_y}"
        terrain = Map.put(state.terrain_map, key, effects["surface_on_throw"])

        # Spread to adjacent if splash radius
        terrain = case effects["surface_radius"] do
          nil -> terrain
          radius ->
            for dy <- -radius..radius, dx <- -radius..radius,
                not (dx == 0 and dy == 0),
                abs(dx) + abs(dy) <= radius,
                reduce: terrain do
              acc ->
                sk = "#{item_target.grid_x + dx},#{item_target.grid_y + dy}"
                Map.put(acc, sk, effects["surface_on_throw"])
            end
        end

        %{state | terrain_map: terrain}
      else
        state
      end

    # Save updated target
    state = %{state | combatants: Map.put(state.combatants, item_target.char_id, item_target)}

    # Consume from inventory (DB side, non-blocking)
    consume_item(actor.char_id, item.id)

    {state, result}
  end

  # ══════════════════════════════════════════════════════════════════
  # LIMIT BREAK
  # ══════════════════════════════════════════════════════════════════

  @doc "Resolve a limit break action"
  def resolve_limit_break(state, actor, target, limit_id, result, _target_limb \\ nil) do
    case load_limit_break(limit_id, actor) do
      nil ->
        {state, %{result | log: ["#{actor.name} can't use that limit break!" | result.log]}}

      limit ->
        cond do
          Map.get(actor, :limitbreak, 0) < 100 ->
            {state, %{result | log: ["Limit break bar not full!" | result.log]}}

          (limit[:break_level] || 0) > Map.get(actor, :breaklevel, 0) ->
            {state, %{result | log: ["Limit break level too low!" | result.log]}}

          (limit[:char_level_req] || 0) > (actor.level || 1) ->
            {state, %{result | log: ["Character level too low for this limit!" | result.log]}}

          true ->
            # Consume limit bar
            actor = Map.put(actor, :limitbreak, 0)

            effects = parse_json(limit[:effects])
            vars = TePhoenix.Battle.Combatant.formula_vars(actor, target || actor)

            log_text = (effects["log"] || "{name} unleashes #{limit.name}!")
                       |> String.replace("{name}", actor.name)

            result = %{result |
              log: ["💥 LIMIT BREAK: #{log_text}" | result.log],
              actions: [%{type: :limit_break, name: limit.name, icon: limit[:icon]} | result.actions]
            }

            # Damage
            {target, result} =
              if effects["damage"] && target do
                formula = get_in(effects, ["damage", "formula"]) || "ATK*4"
                damage = trunc(TePhoenix.Battle.Formula.evaluate(formula, vars))

                # Stance mods
                damage = if actor.stance == "POWER", do: trunc(damage * 1.4), else: damage
                damage = if actor.stance == "MAGIC" and get_in(effects, ["damage", "type"]) == "magic", do: trunc(damage * 1.4), else: damage
                damage = if target.stance == "GUARD", do: trunc(damage * 0.5), else: damage
                damage = max(1, damage)

                t = TePhoenix.Battle.Combatant.apply_damage(target, damage)
                r = %{result |
                  log: ["#{t.name} takes #{damage} damage!" | result.log],
                  actions: [%{type: :limit_damage, target: t.name, amount: damage} | result.actions]
                }
                {t, r}
              else
                {target, result}
              end

            # Heal
            {actor, result} =
              if effects["heal"] do
                formula = get_in(effects, ["heal", "formula"]) || "MHP*0.3"
                heal = trunc(TePhoenix.Battle.Formula.evaluate(formula, vars))
                a = TePhoenix.Battle.Combatant.apply_healing(actor, heal)
                {a, %{result | log: ["#{a.name} recovers #{heal} HP!" | result.log]}}
              else
                {actor, result}
              end

            # Death check
            {target, result} =
              if target && target.current_hp <= 0 do
                {target, %{result |
                  log: ["☠️ #{target.name} has been slain!" | result.log],
                  actions: [%{type: :death, target: target.name} | result.actions]
                }}
              else
                {target, result}
              end

            combatants = Map.put(state.combatants, actor.char_id, actor)
            combatants = if target, do: Map.put(combatants, target.char_id, target), else: combatants
            state = %{state | combatants: combatants}

            # Fire limit_break trigger
            lb_ctx = %{attacker: actor, victim: target, limit_name: limit.name}
            {state, result} = TePhoenix.Battle.Triggers.fire("limit_break_fired", state, lb_ctx, result)

            {state, result}
        end
    end
  end

  # ══════════════════════════════════════════════════════════════════
  # OVERKILL / CHAIN / RATING
  # ══════════════════════════════════════════════════════════════════

  @doc "Calculate overkill multiplier for bonus drops"
  def overkill_multiplier(overkill_damage, settings) do
    if not settings[:enable_overkill_bonus] or overkill_damage <= 0 do
      1.0
    else
      cond do
        overkill_damage >= (settings[:overkill_threshold_large] || 200) -> settings[:overkill_mult_large] || 2.00
        overkill_damage >= (settings[:overkill_threshold_medium] || 100) -> settings[:overkill_mult_medium] || 1.50
        overkill_damage >= (settings[:overkill_threshold_small] || 50) -> settings[:overkill_mult_small] || 1.25
        true -> 1.0
      end
    end
  end

  @doc "Calculate battle rating (S/A/B/C) based on turn count"
  def battle_rating(turn_count, settings) do
    if not settings[:enable_battle_rating] do
      :c
    else
      target = settings[:rating_target_turns] || 8
      cond do
        turn_count <= target -> :s
        turn_count <= target * 1.5 -> :a
        turn_count <= target * 2.5 -> :b
        true -> :c
      end
    end
  end

  @doc "Get XP/gold multiplier for a battle rating"
  def rating_multiplier(rating, settings) do
    case rating do
      :s -> %{xp: settings[:rating_s_xp_mult] || 1.50, gold: settings[:rating_s_gold_mult] || 2.00}
      :a -> %{xp: settings[:rating_a_xp_mult] || 1.25, gold: settings[:rating_a_gold_mult] || 1.50}
      _ -> %{xp: 1.0, gold: 1.0}
    end
  end

  # ── DB helpers ──────────────────────────────────────────────────

  defp load_item(item_id) do
    Repo.one(from i in "game_items", where: i.id == ^item_id,
      select: %{id: i.id, name: i.name, icon: i.icon, effects: i.effects, type: i.type})
  rescue
    _ -> nil
  end

  defp load_limit_break(limit_id, actor) do
    _class_id = Map.get(actor, :class_id)
    query = from l in "game_limit_breaks",
      where: l.id == ^limit_id,
      select: %{id: l.id, name: l.name, icon: l.icon, effects: l.effects,
                break_level: l.break_level, char_level_req: l.char_level_req}

    Repo.one(query)
  rescue
    _ -> nil
  end

  defp consume_item(char_id, item_id) do
    try do
      Repo.query!("UPDATE character_items SET quantity=quantity-1 WHERE character_id=? AND item_id=? AND quantity>0", [char_id, item_id])
      Repo.query!("DELETE FROM character_items WHERE character_id=? AND item_id=? AND quantity<=0", [char_id, item_id])
    rescue
      _ -> nil
    end
  end

  defp parse_json(nil), do: %{}
  defp parse_json(str) when is_binary(str) do
    case Jason.decode(str) do
      {:ok, map} when is_map(map) -> map
      _ -> %{}
    end
  end
  defp parse_json(map) when is_map(map), do: map
  defp parse_json(_), do: %{}
end

File: te.retired/battle/loot.js
// =================================================================
// LOOT SYSTEM — Steal, Drop Tables, Overkill, Chain, Rating
// =================================================================

const { jp } = require('./shared');

// ─── Steal Action ────────────────────────────────────────────────
// Mid-battle steal: roll against target's steal table.
// Two pools: common (high chance) and rare (low chance).
// Steal chance = (attacker.luck + attacker.speed*0.5) vs target level.

async function resolveSteal(db, battle, actor, target, result) {
    if (!battle._settings?.enable_steal) {
        result.log.push('Steal is not available.');
        return result;
    }

    if (target._alreadyStolen) {
        result.log.push(`${target.name} has nothing left to steal!`);
        return result;
    }

    // Prevent stealing from allies
    const actorTeam = battle.getTeamId(actor.charId);
    const targetTeam = battle.getTeamId(target.charId);
    if (actorTeam && targetTeam && actorTeam === targetTeam) {
        result.log.push(`${actor.name} can't steal from an ally!`);
        return result;
    }

    // Load steal table from NPC data
    let stealTable = null;
    try {
        const [rows] = await db.query(
            'SELECT steal_table_json, drop_table_json FROM game_npcs WHERE char_id=? LIMIT 1',
            [target.charId]
        );
        if (rows.length) {
            stealTable = jp(rows[0].steal_table_json, null) || jp(rows[0].drop_table_json, null);
        }
    } catch {}

    if (!stealTable || !stealTable.length) {
        result.log.push(`${target.name} has nothing to steal.`);
        return result;
    }

    // Steal chance: base 50% + (luck * 2) + (speed * 0.5) - (target_level * 3), min 10%, max 95%
    const baseChance = battle._settings.steal_base_chance || 50;
    const chance = Math.min(95, Math.max(10,
        baseChance + (actor.luck || 0) * 2 + (actor.speed || 0) * 0.5 - (target.level || 1) * 3
    ));

    if (Math.random() * 100 > chance) {
        result.log.push(`${actor.name} tries to steal from ${target.name}... but fails! (${Math.round(chance)}% chance)`);
        result.actions.push({ type: 'steal', actor: actor.name, target: target.name, success: false });
        return result;
    }

    // Try rare first, then common
    let stolen = null;
    const rareItems = stealTable.filter(e => e.rarity === 'rare');
    const commonItems = stealTable.filter(e => e.rarity !== 'rare');

    // Rare steal: 20% of the time (if rare items exist)
    if (rareItems.length && Math.random() < (battle._settings.steal_rare_chance || 0.20)) {
        const pick = rareItems[Math.floor(Math.random() * rareItems.length)];
        stolen = pick;
    }

    if (!stolen && commonItems.length) {
        const pick = commonItems[Math.floor(Math.random() * commonItems.length)];
        stolen = pick;
    }

    if (!stolen) {
        result.log.push(`${actor.name} rummages through ${target.name}'s pockets... nothing!`);
        return result;
    }

    // Give item to actor
    try {
        const qty = stolen.qty || 1;
        await db.query(
            'INSERT INTO character_items (character_id, item_id, quantity) VALUES (?,?,?) ON DUPLICATE KEY UPDATE quantity=quantity+VALUES(quantity)',
            [actor.charId, stolen.item_id, qty]
        );
        const [itemRow] = await db.query('SELECT name, icon FROM game_items WHERE id=?', [stolen.item_id]);
        const itemName = itemRow.length ? itemRow[0].name : `Item #${stolen.item_id}`;
        const itemIcon = itemRow.length ? itemRow[0].icon : '📦';

        result.log.push(`${itemIcon} ${actor.name} steals ${itemName}${qty > 1 ? ` x${qty}` : ''} from ${target.name}!`);
        result.actions.push({
            type: 'steal', actor: actor.name, target: target.name, success: true,
            item: { id: stolen.item_id, name: itemName, icon: itemIcon, qty, rare: stolen.rarity === 'rare' }
        });
    } catch (e) {
        result.log.push(`${actor.name} steals something but it crumbles to dust!`);
    }

    target._alreadyStolen = true;
    return result;
}

// ─── Auto Drop on Defeat ─────────────────────────────────────────
// Roll drop table when an enemy dies (not just KO loot action).
// Called from endBattle or checkDeathOrKnockout.

async function rollDropTable(db, enemyCharId, killerCharId, overkillDamage, settings) {
    const drops = [];
    try {
        const [rows] = await db.query(
            'SELECT drop_table_json FROM game_npcs WHERE char_id=? LIMIT 1',
            [enemyCharId]
        );
        if (!rows.length || !rows[0].drop_table_json) return drops;

        const table = jp(rows[0].drop_table_json, []);
        const overkillBonus = settings?.enable_overkill_bonus ? getOverkillMultiplier(overkillDamage, settings) : 1.0;

        for (const entry of table) {
            if (!entry.item_id) continue;
            const adjustedChance = Math.min(100, (entry.chance || 0) * overkillBonus);
            if (Math.random() * 100 > adjustedChance) continue;

            const qty = Math.max(1, entry.min_qty || 1);
            await db.query(
                'INSERT INTO character_items (character_id, item_id, quantity) VALUES (?,?,?) ON DUPLICATE KEY UPDATE quantity=quantity+VALUES(quantity)',
                [killerCharId, entry.item_id, qty]
            );
            const [iRow] = await db.query('SELECT name, icon FROM game_items WHERE id=?', [entry.item_id]);
            drops.push({
                name: iRow.length ? iRow[0].name : `Item #${entry.item_id}`,
                icon: iRow.length ? iRow[0].icon : '📦',
                qty,
            });
        }
    } catch {}
    return drops;
}

// ─── Overkill Bonus ──────────────────────────────────────────────
// If the killing blow exceeds the target's remaining HP by a large margin,
// grant bonus XP/gold and better drop rates.

function getOverkillMultiplier(overkillDamage, settings) {
    if (!settings?.enable_overkill_bonus || !overkillDamage || overkillDamage <= 0) return 1.0;
    // Overkill thresholds: 50% over = small bonus, 100% over = medium, 200% = large
    const pct = overkillDamage; // raw excess damage
    if (pct >= (settings.overkill_threshold_large || 200)) return settings.overkill_mult_large || 2.0;
    if (pct >= (settings.overkill_threshold_medium || 100)) return settings.overkill_mult_medium || 1.5;
    if (pct >= (settings.overkill_threshold_small || 50)) return settings.overkill_mult_small || 1.25;
    return 1.0;
}

function getOverkillXpBonus(overkillDamage, settings) {
    const mult = getOverkillMultiplier(overkillDamage, settings);
    return mult - 1.0; // e.g. 0.5 = +50% XP
}

// ─── Battle Chain ────────────────────────────────────────────────
// Consecutive battles without resting increase drop rates and XP.
// Tracked per-character in memory (resets on rest/inn/death).

function getBattleChainBonus(character, settings) {
    if (!settings?.enable_battle_chain) return { xpMult: 1.0, dropMult: 1.0, chain: 0 };
    const chain = character._battleChain || 0;
    const xpPerChain = settings.chain_xp_bonus || 0.05;
    const dropPerChain = settings.chain_drop_bonus || 0.10;
    const maxChainBonus = settings.chain_max_bonus || 0.50;

    return {
        xpMult: 1.0 + Math.min(maxChainBonus, chain * xpPerChain),
        dropMult: 1.0 + Math.min(maxChainBonus, chain * dropPerChain),
        chain,
    };
}

function incrementBattleChain(character) {
    character._battleChain = (character._battleChain || 0) + 1;
}

function resetBattleChain(character) {
    character._battleChain = 0;
}

// ─── Battle Performance Rating ───────────────────────────────────
// S/A/B/C/D rank based on: turns taken, damage taken, no-death bonus,
// abilities used variety, overkill. Affects XP/gold multiplier.

function calculateBattleRating(battle, winnerTeamId, settings) {
    if (!settings?.enable_battle_rating) return null;

    const members = (battle.teams[winnerTeamId] || [])
        .map(id => battle.combatants[id])
        .filter(c => c && !c.isAI);
    if (!members.length) return null;

    let score = 100; // start at 100, subtract penalties, add bonuses

    // Turns taken: fewer = better
    const turnCount = battle.turnCount || battle.log?.length || 10;
    const targetTurns = settings.rating_target_turns || 8;
    if (turnCount <= targetTurns) score += 20;
    else if (turnCount <= targetTurns * 1.5) score += 0;
    else if (turnCount <= targetTurns * 2) score -= 15;
    else score -= 30;

    // Damage taken: less = better
    const totalMaxHp = members.reduce((s, c) => s + c.maxHp, 0);
    const totalCurrentHp = members.reduce((s, c) => s + Math.max(0, c.currentHp), 0);
    const hpPct = totalCurrentHp / Math.max(1, totalMaxHp);
    if (hpPct >= 0.90) score += 25;       // barely scratched
    else if (hpPct >= 0.70) score += 10;   // took some hits
    else if (hpPct >= 0.40) score -= 5;
    else score -= 20;                       // barely survived

    // No-death bonus
    const anyDied = members.some(c => c.currentHp <= 0 || c._knockedOut);
    if (!anyDied) score += 15;
    else score -= 10;

    // Determine rank
    let rank, label, xpMult, goldMult;
    if (score >= 140) { rank = 'S'; label = 'Perfect'; xpMult = settings.rating_s_xp_mult || 1.50; goldMult = settings.rating_s_gold_mult || 2.00; }
    else if (score >= 110) { rank = 'A'; label = 'Excellent'; xpMult = settings.rating_a_xp_mult || 1.25; goldMult = settings.rating_a_gold_mult || 1.50; }
    else if (score >= 80) { rank = 'B'; label = 'Good'; xpMult = 1.0; goldMult = 1.0; }
    else if (score >= 50) { rank = 'C'; label = 'Average'; xpMult = 0.90; goldMult = 0.90; }
    else { rank = 'D'; label = 'Poor'; xpMult = 0.75; goldMult = 0.75; }

    return { rank, label, score, xpMult, goldMult, turns: turnCount, hpPct: Math.round(hpPct * 100), noDeath: !anyDied };
}

module.exports = {
    resolveSteal,
    rollDropTable,
    getOverkillMultiplier, getOverkillXpBonus,
    getBattleChainBonus, incrementBattleChain, resetBattleChain,
    calculateBattleRating,
};
